File: frontend/src/extension/configSchema.ts
// frontend/src/extension/configSchema.ts
//
// manifest configSchema 的前端读法。后端 pkg/extension/config_schema.go 有同一份契约的
// Go 侧读法（属性名 / 必填项 / format:"password" 字段），两侧都只读 manifest，不各自
// 引入额外约定。

export interface ExtensionConfigProperty {
  type?: string;
  format?: string;
  enum?: string[];
  title?: string;
  description?: string;
  placeholder?: string;
}

export interface ExtensionConfigSchema {
  type?: string;
  properties?: Record<string, ExtensionConfigProperty>;
  required?: string[];
  propertyOrder?: string[];
}

/** 返回 format:"password" 的属性名——它们保存前要经后端加密，展示时要打码。 */
export function passwordFields(schema?: ExtensionConfigSchema): string[] {
  const props = schema?.properties ?? {};
  return Object.entries(props)
    .filter(([, prop]) => prop.format === "password")
    .map(([name]) => name);
}

File: pkg/extension/config_schema.go
package extension

import "sort"

// PasswordFieldsFromSchema extracts property names that have "format": "password"
// from a JSON Schema configSchema.
func PasswordFieldsFromSchema(schema map[string]any) []string {
	if len(schema) == 0 {
		return nil
	}
	props, ok := schema["properties"].(map[string]any)
	if !ok {
		return nil
	}
	var fields []string
	for name, v := range props {
		prop, ok := v.(map[string]any)
		if !ok {
			continue
		}
		if fmt, ok := prop["format"].(string); ok && fmt == "password" {
			fields = append(fields, name)
		}
	}
	sort.Strings(fields)
	return fields
}

// ConfigSchemaProperties returns the declared property names of a configSchema,
// sorted. Empty for a schema without a properties object.
func ConfigSchemaProperties(schema map[string]any) []string {
	props, ok := schema["properties"].(map[string]any)
	if !ok {
		return nil
	}
	names := make([]string, 0, len(props))
	for name := range props {
		names = append(names, name)
	}
	sort.Strings(names)
	return names
}

// ConfigSchemaRequired returns the property names listed in a configSchema's
// "required" array, sorted. Entries that are not declared properties are dropped:
// a required name nothing declares can never be supplied.
func ConfigSchemaRequired(schema map[string]any) []string {
	raw, ok := schema["required"].([]any)
	if !ok {
		return nil
	}
	props, _ := schema["properties"].(map[string]any)
	names := make([]string, 0, len(raw))
	for _, item := range raw {
		name, ok := item.(string)
		if !ok {
			continue
		}
		if _, declared := props[name]; !declared {
			continue
		}
		names = append(names, name)
	}
	sort.Strings(names)
	return names
}
